macaroon.js

```diff
@@ -338,197 +392,249 @@ const readFieldV2Optional = function(buf, maybeFieldType) {
+        _identifierBits: bytesToBits(identifierBytes),
+        _vidBits: cav.vidBytes && bytesToBits(requireBytes(cav.vidBytes, 'Verification ID')),
+      };
+    }) : [];
   }
 
-  get location() {
-    return this._location;
+  /**
+   * Return the caveats associated with the macaroon,
+   * as an array of caveats. A caveat is represented
+   * as an object with an identifier field (Uint8Array)
+   * and (for third party caveats) a location field (string),
+   * and verification id (Uint8Array).
+   * @return {Array} The macaroon's caveats.
+   */
+  get caveats() {
+    if (!this._caveats) {
+      return [];
+    }
+    return this._caveats.map(cav => {
+      return cav._vidBits ? {
+        identifier: bitsToBytes(cav._identifierBits),
+        location: cav._locationStr,
+        vid: bitsToBytes(cav._vidBits),
+      } : {
+        identifier: bitsToBytes(cav._identifierBits),
+      };
+    });
   }
 
-  get identifier() {
-    return this._identifier;
+  /**
+   * Return the location of the macaroon.
+   * @return {string} The macaroon's location.
+   */
+  get location() {
+    return this._locationStr;
   }
 
-  get signature() {
-    return bitArrayToUint8Array(this._signature);
+  /**
+   * Return the macaroon's identifier.
+   * @return {Uint8Array} The macaroon's identifier.
+   */
+  get identifier() {
+    return bitsToBytes(this._identifierBits);
   }
 
   /**
-    Adds a first or third party caveat.
-    @param {String} caveatId
-    @param {bitArray} verificationId For a first party caveat, must be null
-      otherwise must be bitArray.
-    @param {String} location For a first party caveat, must be null otherwise
-      must be String.
-  */
-  addCaveat(caveatId, verificationId, location) {
-    const caveat = {
-      _identifier: requireString(caveatId, 'Macaroon caveat id'),
-      _vid: null,
-      _location: null,
-    };
-    if (verificationId) {
-      caveat._location = requireString(location, 'Macaroon caveat location');
-      caveat._vid = bytesToBitArray(
-        requireUint8Array(verificationId, 'Macaroon caveat verification id'));
-    }
-    this._caveats.push(caveat);
-    this._signature = keyedHash2(
-      this._signature, caveat._vid, sjcl.codec.utf8String.toBits(caveatId));
+   * Return the signature of the macaroon.
+   * @return {string} The macaroon's signature.
+   */
+  get signature() {
+    return bitsToBytes(this._signatureBits);
   }
 
   /**
     Adds a third party caveat to the macaroon. Using the given shared root key,
     caveat id and location hint. The caveat id should encode the root key in
     some way, either by encrypting it with a key known to the third party or by
     holding a reference to it stored in the third party's storage.
-    @param {bitArray} rootKey
-    @param {String} caveatId
-    @param {String} location
+    @param {Uint8Array} rootKeyBytes
+    @param {Uint8Array | string} caveatIdBytes
+    @param {optional String} locationStr
```

Comment:
in JSDoc optional params are done in the following way:
`@param {String} [locationStr]`

Reply:
done